### ISSUE TITLE
Unexpected Result By Index

### ISSUE THREAD

**Issue by bajinsheng** [pingcap/tidb#46572]
### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE t0(c0 FLOAT);
INSERT INTO t0 VALUES (0.9), (0.8);
CREATE INDEX i0 ON t0(c0);
SET @@sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';


SELECT t0.c0 FROM t0 GROUP BY CAST(t0.c0 AS DECIMAL); -- {0.8}
SELECT /*+ IGNORE_INDEX(t0, i0)*/t0.c0 FROM t0 GROUP BY CAST(t0.c0 AS DECIMAL); -- {0.9}
```

### 3. What did you see instead (Required)
Two queries return different results.

### 4. What is your TiDB version? (Required)

| Release Version: v7.4.0-alpha-239-g4f2f5e1061
Edition: Community
Git Commit Hash: 4f2f5e1061ddc55a2dd349e1df67847c0edd1f23
Git Branch: master
UTC Build Time: 2023-08-30 12:06:00
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: unistore |

**Comment by qw4990** [comment]
The root cause is that this query breaks the `only_full_group_by` attribute and then causes the result unstable.
First, values of `c0` on the index `i0` and `table` have different order:
```
mysql> select c0 from t0 ignore index(i0);
+------+
| c0   |
+------+
|  0.9 |
|  0.8 |
+------+
2 rows in set (0.00 sec)

mysql> select c0 from t0 ignore index(i0);
+------+
| c0   |
+------+
|  0.9 |
|  0.8 |
+------+
2 rows in set (0.00 sec)
```

And then the `CAST` predicate converts both `0.9` and `0.8` to `1`, so these 2 rows will be in the same group:
```
mysql> select cast(t0.c0 as decimal) from t0;
+------------------------+
| cast(t0.c0 as decimal) |
+------------------------+
|                      1 |
|                      1 |
+------------------------+
```

And finally, since the query breaks the `only_full_group_by` attribute, any of those 2 rows can be returned as the result:
![image](https://github.com/pingcap/tidb/assets/7499936/e43c4726-303d-41ea-8384-13002449579b)

**Comment by mjonss** [comment]
Closing, already marked as wontfix, but the explanation also means it is expected behavior, since only_full_group_by is not set, and there are no order in the query.

So unstable results should be given, similar to `select * from t limit 1` does not guarantee stable result, even if MySQL normally returns a stable result. TiDB is a distributed database and retrieves rows in parallel, so without explicit order, such result is not guaranteed to be stable / rows not returned in same order.
